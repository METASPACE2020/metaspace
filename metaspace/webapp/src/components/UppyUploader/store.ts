// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from 'vue'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State | any>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach((listener) => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
