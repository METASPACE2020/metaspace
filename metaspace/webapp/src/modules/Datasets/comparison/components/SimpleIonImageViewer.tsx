import { computed, defineComponent, onMounted, reactive, ref, watch } from '@vue/composition-api'
import './SimpleIonImageViewer.scss'
import IonImageViewer from '../../../../components/IonImageViewer'
import fitImageToArea, { FitImageToAreaResult } from '../../../../lib/fitImageToArea'
import { ImagePosition } from '../../../ImageViewer/ionImageState'
import { loadPngFromUrl, processIonImage, renderScaleBar } from '../../../../lib/ionImageRendering'
import config from '../../../../lib/config'
import createColormap from '../../../../lib/createColormap'
import Vue from 'vue'
import safeJsonParse from '../../../../lib/safeJsonParse'
import ImageSaver from '../../../ImageViewer/ImageSaver.vue'
import { Input } from '../../../../lib/element-ui'

interface SimpleIonImageViewerProps {
  isActive: boolean
  resetViewPort: boolean
  isNormalized: boolean
  showOpticalImage: boolean
  normalizationData: any
  width: number
  height: number
  annotations: any[]
  scaleType: string
  scaleBarColor: string
  colormap: string
}

interface ImageSettings {
  intensity: any
  ionImagePng: any
  pixelSizeX: number
  pixelSizeY: number
  ionImageLayers: any
  imageFit: Readonly<FitImageToAreaResult>
  lockedIntensities: [number | undefined, number | undefined]
  annotImageOpacity: number
  opticalOpacity: number
  imagePosition: ImagePosition,
  pixelAspectRatio: number
  imageZoom: number
  showOpticalImage: boolean
  userScaling: [number, number],
  imageScaledScaling: [number, number],
  scaleBarUrl: Readonly<string[]>,
}

interface SimpleIonImageViewerState {
  mode: string
  imageSettings: ImageSettings | any,
  colorSettings: any
  ionImagePng: any
}

const channels: any = {
  magenta: 'rgb(255, 0, 255)',
  green: 'rgb(0, 255, 0)',
  blue: 'rgb(0, 0, 255)',
  red: 'rgb(255, 0, 0)',
  yellow: 'rgb(255, 255, 0)',
  cyan: 'rgb(0, 255, 255)',
  orange: 'rgb(255, 128, 0)',
  violet: 'rgb(128, 0, 255)',
  white: 'rgb(255, 255, 255)',
}

export const SimpleIonImageViewer = defineComponent<SimpleIonImageViewerProps>({
  name: 'SimpleIonImageViewer',
  props: {
    annotations: { type: Array },
    isActive: { type: Boolean, required: false, default: false },
    width: { type: Number, required: false },
    height: { type: Number, required: false },
    colormap: {
      type: String,
      default: 'Viridis',
    },
    scaleType: {
      type: String,
      default: 'linear',
    },
    scaleBarColor: {
      type: String,
      default: '#000000',
    },
    showOpticalImage: {
      type: Boolean,
      default: true,
    },
    isNormalized: {
      type: Boolean,
      default: false,
    },
    resetViewPort: {
      type: Boolean,
      default: false,
    },
    normalizationData: {
      type: Object,
      default: null,
    },
  },
  setup(props, { emit }) {
    const state = reactive<SimpleIonImageViewerState>({
      mode: props.isActive ? 'MULTI' : 'SINGLE',
      imageSettings: {},
      colorSettings: {},
      ionImagePng: null,
    })

    const container = ref(null)

    onMounted(() => {
      console.log('start')
      startImageSettings()
    })

    // reset view port globally
    watch(() => props.resetViewPort, (newValue) => {
      if (newValue) {
        emit('resetViewPort', false)
        resetViewPort()
      }
    })

    const resetViewPort = () => {
      if (state.imageSettings) {
        state.imageSettings!.imagePosition = defaultImagePosition()
      }
    }

    const getMetadata = (annotation: any) => {
      const datasetMetadataExternals = {
        Submitter: annotation.dataset.submitter,
        PI: annotation.dataset.principalInvestigator,
        Group: annotation.dataset.group,
        Projects: annotation.dataset.projects,
      }
      return Object.assign(safeJsonParse(annotation.dataset.metadataJson), datasetMetadataExternals)
    }

    const ionImage = (ionImagePng: any, isotopeImage: any,
      scaleType: any = 'linear', userScaling: any = [0, 1], normalizedData: any = null) => {
      if (!isotopeImage || !ionImagePng) {
        return null
      }
      const { minIntensity, maxIntensity } = isotopeImage
      return processIonImage(ionImagePng, minIntensity, maxIntensity, scaleType
        , userScaling, undefined, normalizedData)
    }

    const ionImageLayers = () => {
      const { annotations, isNormalized, normalizationData } = props
      const { imageSettings, colorSettings } = state

      if (annotations === null || annotations[0] === null || imageSettings == null || colorSettings === null) {
        return []
      }

      const ionImages = []

      for (let index = 0; index < annotations?.length; index++) {
        const annotation = annotations[index]
        const ionImagePng = state.ionImagePng[index]

        const finalImage = ionImage(ionImagePng,
          annotation.isotopeImages[0],
          props.scaleType, imageSettings.imageScaledScaling,
          isNormalized && normalizationData
            ? normalizationData : null)
        const hasOpticalImage = annotation.dataset.opticalImages[0]?.url !== undefined

        if (finalImage) {
          ionImages.push({
            ionImage: finalImage,
            colorMap: createColormap(colorSettings[annotation.id].value,
              hasOpticalImage && props.showOpticalImage
                ? 'linear' : 'constant',
              hasOpticalImage && index === 0 && props.showOpticalImage
                ? (imageSettings.annotImageOpacity || 1) : 1),
          })
        }
      }

      return ionImages.filter((a: any) => a !== null)
    }

    const scaleBars = () => {
      const { annotations } = props
      const { imageSettings, colorSettings } = state

      return annotations.map((annotation: any, index: number) => {
        return renderScaleBar(
          imageSettings.ionImageLayers[index]?.ionImage,
          createColormap(props.colormap),
          true,
        )
      })
    }

    const imageFit = () => {
      const { imageSettings } = state
      const { width = props.width, height = props.height } = imageSettings?.ionImagePng || {}

      return fitImageToArea({
        imageWidth: width,
        imageHeight: height / (imageSettings?.pixelAspectRatio || 1),
        areaWidth: props.width,
        areaHeight: props.height,
      })
    }

    const defaultImagePosition = () => ({
      // This is a function so that it always makes a separate copy for each image
      zoom: 1,
      xOffset: 0,
      yOffset: 0,
    })

    const startImageSettings = async() => {
      const { annotations, isActive } = props
      const annotation = annotations[0]
      const ionImagesPng = []

      const metadata = getMetadata(annotation)
      // eslint-disable-next-line camelcase
      const pixelSizeX = metadata?.MS_Analysis?.Pixel_Size?.Xaxis || 0
      // eslint-disable-next-line camelcase
      const pixelSizeY = metadata?.MS_Analysis?.Pixel_Size?.Yaxis || 0

      for (let i = 0; i < annotations?.length; i++) {
        const annotationItem = annotations[i]
        state.colorSettings[annotationItem.id] = computed(() => isActive
          ? props.colormap : Object.keys(channels)[i % Object.keys(channels).length])
        const ionImagePng = await loadPngFromUrl(annotationItem.isotopeImages[0].url)
        ionImagesPng.push(ionImagePng)
      }

      state.ionImagePng = ionImagesPng

      const imageSettings : ImageSettings = reactive({
        intensity: null, // @ts-ignore // Gets set later, because ionImageLayers needs state.gridState[key] set
        ionImagePng: ionImagesPng[0],
        pixelSizeX,
        pixelSizeY,
        // ionImageLayers and imageFit rely on state.gridState[key] to be correctly set - avoid evaluating them
        // until this has been inserted into state.gridState
        ionImageLayers: computed(() => ionImageLayers()),
        imageFit: computed(() => imageFit()),
        lockedIntensities: [undefined, undefined],
        annotImageOpacity: 1.0,
        opticalOpacity: 1.0,
        imagePosition: defaultImagePosition(),
        pixelAspectRatio:
            config.features.ignore_pixel_aspect_ratio ? 1
              : pixelSizeX && pixelSizeY && pixelSizeX / pixelSizeY || 1,
        imageZoom: 1,
        showOpticalImage: props.showOpticalImage,
        userScaling: [0, 1],
        imageScaledScaling: [0, 1],
        scaleBarUrl: computed(() => scaleBars()),
      })
      Vue.set(state, 'imageSettings', imageSettings)
    }

    const handleImageMove = ({ zoom, xOffset, yOffset }: any) => {
      const gridCell = state.imageSettings
      if (gridCell != null) {
        gridCell.imagePosition.zoom = zoom / gridCell.imageFit.imageZoom
        gridCell.imagePosition.xOffset = xOffset
        gridCell.imagePosition.yOffset = yOffset
      }
    }

    return () => {
      const { width, height, annotations } = props
      const { imageSettings } = state

      if (!imageSettings || !imageSettings.ionImageLayers
        || imageSettings.ionImageLayers.length === 0 || !annotations) {
        return null
      }

      return (
        <div
          ref={container}
          class={'ds-simple-ion-image-container relative'}
          style={{
            width,
            height,
          }}
        >
          <IonImageViewer
            height={height}
            width={width}
            zoom={imageSettings!.imagePosition?.zoom
              * imageSettings!.imageFit?.imageZoom}
            xOffset={imageSettings!.imagePosition?.xOffset || 0}
            yOffset={imageSettings!.imagePosition?.yOffset || 0}
            isLoading={false}
            ionImageLayers={imageSettings!.ionImageLayers}
            scaleBarColor={props.scaleBarColor}
            scaleType={props.scaleType}
            pixelSizeX={imageSettings!.pixelSizeX}
            pixelSizeY={imageSettings!.pixelSizeY}
            pixelAspectRatio={imageSettings!.pixelAspectRatio}
            opticalOpacity={imageSettings!.opticalOpacity}
            imageHeight={imageSettings!.ionImageLayers[0]?.ionImage?.height }
            imageWidth={imageSettings!.ionImageLayers[0]?.ionImage?.width }
            minZoom={imageSettings!.imageFit.imageZoom / 4}
            maxZoom={imageSettings!.imageFit.imageZoom * 20}
            opticalSrc={props.showOpticalImage
              ? annotations[0]?.dataset?.opticalImages[0]?.url
              : undefined}
            opticalTransform={props.showOpticalImage
              ? annotations[0]?.dataset?.opticalImages[0]?.transform
              : undefined}
            scrollBlock
            showPixelIntensity
            onMove={handleImageMove}
          />
          <ImageSaver
            class="absolute top-0 left-0 mt-3 ml-3 dom-to-image-hidden"
            domNode={container.value}
          />
        </div>
      )
    }
  },
})
