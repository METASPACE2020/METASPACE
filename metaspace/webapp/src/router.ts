/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage.vue'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DesignPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DesignPage" */ './design/DesignPage.vue'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/GroupProfile/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/GroupsListPage.vue'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage.vue'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/GroupProfile/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },
    { path: '/admin/groups', component: asyncPages.GroupsListPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignPage, meta: { footer: true, flex: true } },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
