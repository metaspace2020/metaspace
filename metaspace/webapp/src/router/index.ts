/* eslint-disable vue/max-len */
import { createRouter, createWebHistory } from 'vue-router'
import { Component } from 'vue'
import AboutPage from '../modules/App/AboutPage'
import DatasetsPage from '../modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from '../modules/Account'
import { redirectAfterSignIn } from '../modules/Account/signInReturnUrl'
import { updateDBParam } from '../modules/Filters/url'
import NotFound from '../modules/App/NotFoundPage.vue'

const asyncPagesFreelyTyped = {
  AnnotationsPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ '../modules/Annotations/AnnotationsPage.vue'
    ),
  DatasetSummary: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ '../modules/Datasets/summary/DatasetSummary.vue'
    ),
  MetadataEditPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ '../modules/MetadataEditor/MetadataEditPage.vue'
    ),
  ImageAlignmentPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ '../modules/ImageAlignment/ImageAlignmentPage.vue'
    ),
  UploadPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ '../modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ '../modules/Datasets/overview/DatasetOverviewPage'
    ),
  DatasetComparisonPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetComparisonPage" */ '../modules/Datasets/comparison/DatasetComparisonPage'
    ),
  DatasetBrowserPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetBrowserPage" */ '../modules/Datasets/imzml/DatasetBrowserPage'
    ),
  DatasetEnrichmentPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "DatasetEnrichmentPage" */ '../modules/Datasets/enrichment/DatasetEnrichmentPage'
    ),
  SpottingProjectPage: () =>
    import(
      /* webpackPrefetch: true, webpackChunkName: "SpottingProjectPage" */ '../modules/SpottingProject/DashboardPage'
    ),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/HelpPage.vue'),
  EditUserPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/TermsPage.vue'),
  PublicationsPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ '../modules/App/PublicationsPage'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () =>
    import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ '../modules/Project/ViewProjectPage.vue'),

  // Separate bundle for design docs
  DesignTOC: () => import(/* webpackChunkName: "DesignBundle" */ '../design/TOCPage.vue'),
  DesignStyleGuide: () => import(/* webpackChunkName: "DesignBundle" */ '../design/StyleGuidePage.vue'),
  DesignIcons: () => import(/* webpackChunkName: "DesignBundle" */ '../design/IconsPage.vue'),
  DesignComponents: () => import(/* webpackChunkName: "DesignBundle" */ '../design/ComponentsPage.vue'),
  DesignForms: () => import(/* webpackChunkName: "DesignBundle" */ '../design/FormsPage.vue'),
  ContactPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ContactPage" */ '../modules/Contact/Contact'),
}

const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, Component>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

export const routes: any = [
  { path: '/', name: 'home', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/about', name: 'about', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
  { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
  { path: '/contact', name: 'contact', component: asyncPages.ContactPage },
  {
    path: '/datasets',
    name: 'dataset-list',
    component: DatasetsPage,
    children: [
      { path: '', name: 'datasets', component: asyncPages.DatasetTable },
      { path: 'summary', name: 'summary', component: asyncPages.DatasetSummary },
    ],
  },
  {
    path: '/datasets/:dataset_id/comparison',
    name: 'datasets-comparison',
    component: asyncPages.DatasetComparisonPage,
  },
  { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
  {
    path: '/datasets/:dataset_id/add-optical-image',
    name: 'add-optical-image',
    component: asyncPages.ImageAlignmentPage,
  },
  { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
  { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
  { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
  { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetEnrichmentPage },
  { path: '/upload', name: 'upload', component: asyncPages.UploadPage },
  { path: '/help', name: 'help', component: asyncPages.HelpPage, meta: { footer: true } },
  { path: '/user/me', name: 'profile', component: asyncPages.EditUserPage },

  { path: '/admin/health', name: 'admin', component: asyncPages.SystemHealthPage },

  { path: '/account/sign-in', name: 'sign-in', component: DialogPage, props: { dialog: 'signIn' } },
  { path: '/account/sign-in-success', name: 'sign-in-success', redirect: redirectAfterSignIn },
  { path: '/account/create-account', name: 'create-acc', component: DialogPage, props: { dialog: 'createAccount' } },
  { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
  { path: '/account/reset-password', component: ResetPasswordPage },

  { path: '/groups', name: 'group-list', component: asyncPages.GroupsListPage },
  { path: '/group/create', name: 'group-create', component: asyncPages.CreateGroupPage },
  { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

  { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
  {
    // Legacy URL sent in "request access" emails up until Feb 2019
    path: '/project/:projectIdOrSlug/manage',
    redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } },
  },
  { path: '/projects', name: 'project-list', component: asyncPages.ProjectsListPage },

  { path: '/terms', name: 'terms', component: asyncPages.TermsPage, meta: { footer: true } },
  { path: '/privacy', name: 'privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
  { path: '/publications', name: 'publication-list', component: asyncPages.PublicationsPage, meta: { footer: true } },

  { path: '/design', name: 'design', component: asyncPages.DesignTOC, meta: { footer: true, flex: true } },
  { path: '/design/styleguide', component: asyncPages.DesignStyleGuide, meta: { footer: true, flex: true } },
  { path: '/design/icons', component: asyncPages.DesignIcons, meta: { footer: true, flex: true } },
  { path: '/design/components', component: asyncPages.DesignComponents, meta: { footer: true, flex: true } },
  { path: '/design/forms', component: asyncPages.DesignForms, meta: { footer: true, flex: true } },

  { path: '/detectability', name: 'detectability', component: asyncPages.SpottingProjectPage },

  { path: '/:pathMatch(.*)*', name: 'not-found', component: NotFound, meta: { footer: true, flex: true } },
]

const router = createRouter({
  // @ts-ignore
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
  scrollBehavior(to, from, savedPosition) {
    // return desired position
    if (savedPosition) {
      return savedPosition
    }
    return { top: 0 }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if ((to.path !== from.path && daysSincePageLoad > 1.75) || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
