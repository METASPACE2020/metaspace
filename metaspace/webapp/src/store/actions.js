import { fetchOptionListsQuery } from '../api/metadata'
import { decodeParams } from '../modules/Filters'
import { computed } from 'vue'
import store from '../store/index'

export default {
  async initFilterLists(context, apolloClient) {
    if (context.state.filterListsLoading || context.state.filterLists != null) return
    await context.commit('setFilterListsLoading')

    const response = await apolloClient.query({
      query: fetchOptionListsQuery,
      fetchPolicy: 'cache-first',
    })

    // set annotationIds default according to values passed from snapshot to the store state
    context.commit('setFilterLists', {
      ...response.data,
      annotationIds: computed(() => store.state.snapshotAnnotationIds),
    })

    // Refresh the current filter so that computed defaults that depend on `filterLists` are applied
    const filter = decodeParams(context.state.route, context.state.filterLists)

    context.commit('updateFilter', filter)
  },
}
