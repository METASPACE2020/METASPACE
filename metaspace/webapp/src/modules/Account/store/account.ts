import { Module } from 'vuex';
import { DialogType } from '../dialogs';
import router from '../../../router';

export interface AccountState {
  dialog: DialogType | null;
  showDialogAsPage: boolean;
}

const account = {
  namespaced: true,
  state: {
    dialog: null,
    showDialogAsPage: false
  },
  mutations: {
    showDialog(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = false;
    },
    showDialogAsPage(state: AccountState, dialog: DialogType) {
      state.dialog = dialog;
      state.showDialogAsPage = true;
    },
    hideDialog(state: AccountState, dialog: DialogType) {
      if (state.dialog === dialog) {
        state.dialog = null;
        if (state.showDialogAsPage) {
          router.push('/');
        }
      }
    }
  }
};

export default account as Module<AccountState, any>;
