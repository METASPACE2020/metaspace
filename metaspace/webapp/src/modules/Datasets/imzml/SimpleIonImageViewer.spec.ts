import { mount } from '@vue/test-utils'
import Vue from 'vue'
import SimpleIonImageViewer from './SimpleIonImageViewer'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'
import store from '../../../store'
import router from '../../../router'

describe('SimpleIonImageViewer', () => {
  const propsData = {
    annotation: {
      id: '2021-05-10_16h11m35s_40994',
      sumFormula: 'C34H46O11',
      adduct: '+H',
      ion: 'C34H46O11+H+',
      ionFormula: 'C34H47O11',
      database: 'ChEBI',
      msmScore: 0.00000276404,
      rhoSpatial: 0.00000313589,
      rhoSpectral: 0.881923,
      rhoChaos: 0.99943,
      fdrLevel: 0.5,
      mz: 631.3112495016694,
      colocalizationCoeff: null,
      offSample: null,
      offSampleProb: null,
      isotopeImages: [
        {
          mz: 631.3112495016694,
          url: 'blob:http://localhost:8999/8e5542e8-2f21-40fc-8562-152c627f2a91',
          minIntensity: 0,
          maxIntensity: 4848650,
          totalIntensity: 66381492,
        },
      ],
      isomers: [],
      isobars: [],
      countPossibleCompounds: 1,
      possibleCompounds: [
        {
          name: 'ajugatakasin A',
          imageURL: '/mol-images/ChEBI/CHEBI:69876.svg',
          information: [
            {
              database: 'ChEBI',
              url: 'nullCHEBI:69876',
            },
          ],
        },
      ],
    },
    ionImageUrl: 'blob:http://localhost:8999/8e5542e8-2f21-40fc-8562-152c627f2a91',
    pixelSizeX: 100,
    pixelSizeY: 100,
  }
  const testHarness = Vue.extend({
    components: {
      SimpleIonImageViewer,
    },
    render(h) {
      return h(SimpleIonImageViewer, { props: this.$attrs })
    },
  })

  beforeAll(() => {
    Vue.use(Vuex)
    sync(store, router)
  })

  it('it should match snapshot when empty', async() => {
    const wrapper = mount(testHarness, {
      store,
      router,
      propsData: {
        isEmpty: true,
        isLoading: false,
      },
    })
    await Vue.nextTick()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('it should match snapshot when loading chart data', async() => {
    const wrapper = mount(testHarness, {
      store,
      router,
      propsData,
    })
    await Vue.nextTick()

    expect(wrapper.element).toMatchSnapshot()
  })
})
