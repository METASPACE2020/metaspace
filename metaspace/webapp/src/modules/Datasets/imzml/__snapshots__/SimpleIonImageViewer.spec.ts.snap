// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleIonImageViewer it should match snapshot when empty 1`] = `
<div
  class="simple-ion-image-container"
  isempty="true"
/>
`;

exports[`SimpleIonImageViewer it should match snapshot when loading chart data 1`] = `
<div
  class="simple-ion-image-container"
  isdataloading="true"
/>
`;
